### PR TITLE
Implement quorum reads and response merging for /v1/alerts.

### PR DESCRIPTION
**Pull request by stevesg**
**What this PR does**:

Reading alert listings via /v1/alerts will now read from a quorum of
replicas and return a merged response. The merging logic returns a union
of the alerts from all responses. When the same alert is returned by
multiple replicas, one is just chosen arbitrarily. This should be
sufficient as alerts posted to each replica should be identical.

The replicas may have some disagreement on the alert status, but this
should become consistent over time. For the /v2 API, we can use the
"updatedAt" timestamp to choose the most up-to-date status, but this
field is not available for the /v1 API.

Merging for /v2/alerts and /v2/alerts/groups to follow.

**Checklist**
- [X] Tests updated
- [X] ~~Documentation added~~
- [X] ~~`CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`~~

### CODE REVIEW

File: pkg/alertmanager/distributor.go
Comment (pracucci) [inline pkg/alertmanager/distributor.go:316]: [nit]

```suggestion
	// have already been checked for success or failure, and that the headers will always
```

File: integration/alertmanager_test.go
Comment (pracucci) [inline integration/alertmanager_test.go:502]: [nit] Just to clarify

```suggestion
			// Endpoint: GET /v1/alerts
```

File: pkg/alertmanager/merger/merger.go
Comment (pracucci) [inline pkg/alertmanager/merger/merger.go:14]: [nit] I understand this can't happen right now, but to be 100% safe maybe we should add an:

```
if len(in) == 0 {
  return nil, nil
}
```

**Review by pracucci — APPROVED** [review]
Amazing job, flawless! You 🤘 !

_I just left few non-blocking nits._

File: pkg/alertmanager/merger/v1_alerts.go
Comment (pstibrany) [inline pkg/alertmanager/merger/v1_alerts.go:33]: If all responses have status not-success, should we pick one and return it? It could be better than returning 5xx.

File: pkg/alertmanager/merger/v1_alerts.go
Comment (pstibrany) [inline pkg/alertmanager/merger/v1_alerts.go:48]: super-nit: we can return result of `json.Marshal` directly

File: pkg/alertmanager/merger/merger.go
Comment (pstibrany) [inline pkg/alertmanager/merger/merger.go:1]: Super-Nit: Not sure this package carries its own weight. I know there will be v2 implementation too, but it's still looks pretty thin. Same goes for `Merger` interface.

**Review by pstibrany — APPROVED** [review]
LGTM, thank you!

File: pkg/alertmanager/merger/v1_alerts.go
Comment (stevesg) [reply pkg/alertmanager/merger/v1_alerts.go:33]: My understanding of the distributor is that only successful responses get this far as they are checked earlier on. https://github.com/cortexproject/cortex/blob/master/pkg/alertmanager/distributor.go#L171

File: pkg/alertmanager/merger/merger.go
Comment (stevesg) [reply pkg/alertmanager/merger/merger.go:1]: I agree and wasn't 100% sure what to do. The `pkg/alertmanager` package is getting very busy, was my reasoning here.

Would a `pkg/alertmanager/distributor` perhaps be better?

Also regarding the interface, I'd be equally happy with a functions, if that's what you meant? (I was mainly taking inspiration from here https://github.com/cortexproject/cortex/blob/94b0137e7133652f32a6ca641510b8c4c861791e/pkg/querier/queryrange/query_range.go#L67)

Comment (pstibrany) [reply pkg/alertmanager/merger/merger.go:1]: Personally I would just use `pkg/alertmanager`, I think it's still a small package.

> Also regarding the interface, I'd be equally happy with functions, if that's what you meant?

Yes, I was thinking about just using functions here. My reasoning is that 1) we're only going to have 2 implementations (v1, v2 api) and 2) client doesn't need to supply its own implementation. But I am first to admit this is just a personal preference for this specific case, and don't mind keeping it as-is.

Comment (pracucci) [reply pkg/alertmanager/merger/merger.go:1]: I don't have a strong opinion here. I personally did like the `merger` package, given it's content is self contained but moving to `pkg/alertmanager` looks good to me as well. Let's just not block on this. V2 merging PR is awaiting for this PR! 😉 

> Would a pkg/alertmanager/distributor perhaps be better?

I wouldn't name the package distributor because we already have `pkg/distributor` and in the past (when we tried `pkg/alertmanager/distributor` at some point) we realised it was causing much confusion when reading the code.

Comment (stevesg) [reply pkg/alertmanager/merger/merger.go:1]: Leaving as-is to unblock, happy to refactor after the fact.

**Pull request merged**
